changefeedccl: add helper to verify testfeed ordering guarantees

spiffyy99 opened the pull request:
Previously, the ordering of payloads was not considered. This was sufficient until concurrency was introduced into testfeeds, making some assertions on unordered payloads unreliable. This new function ensures payloads from testfeeds satisfy CDC ordering guarantees.

Resolves #67394

Release note: None

Epic CRDB-7748

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67499)
<!-- Reviewable:end -->

spiffyyeng commented on pkg/ccl/changefeedccl/helpers_test.go:
```diff
@@ -62,8 +63,8 @@ func waitForSchemaChange(
-func readNextMessages(f cdctest.TestFeed, numMessages int, stripTs bool) ([]string, error) {
-	var actual []string
+func readNextMessages(f cdctest.TestFeed, numMessages int) ([]cdctest.TestFeedMessage, error) {
+	var actual []cdctest.TestFeedMessage
```
return as `TestFeedMessage` here because we need messages in struct format to verify ordering guarantees

spiffyyeng commented on pkg/ccl/changefeedccl/helpers_test.go:
```diff
@@ -80,53 +81,137 @@ func readNextMessages(f cdctest.TestFeed, numMessages int, stripTs bool) ([]stri
+	}
+
+	// strip timestamps after checking per-key ordering since check uses timestamps
+	if stripTs {
```
move `stripTs` logic here since we might need to check per-key ordering first (which needs to be done before stripping the timestamps)

spiffyy99 committed:
changefeedccl: add helper to verify testfeed ordering guarantees

Previously, the ordering of payloads was not considered. This was
sufficient until concurrency was introduced into testfeeds, making
some assertions on unordered payloads invalid. This new function
ensures payloads from testfeeds satisfy CDC ordering guarantees.

Resolves #67394

Release note: None

HonoreDB reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67499)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng and @stevendanna)

___
*[pkg/ccl/changefeedccl/helpers_test.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67499#-MeR8I1H-gA69cU79ajl:-MeR8I1H-gA69cU79ajm:b-2gw5ae) ([raw file](https://github.com/cockroachdb/cockroach/blob/af6e2ebccd0737cb2f93e9bdfabbdd6454796b3a/pkg/ccl/changefeedccl/helpers_test.go#L131)):*
> ```Go
> func checkPerKeyOrdering(payloads []cdctest.TestFeedMessage) (bool, error) {
> 	// map key to list of timestamp, ensure each list is ordered
> 	keysToTimestamps := make(map[string][]float64)
> ```

nit: this can just be a `map[string]float64` since you only need to see the last timestamp for a key to check the ordering guarantee.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
Cool. Did this survive with the webhook changes under `make stress`?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67499#-:-MeUlDGW1qFw2B-eZG2H:b1zzdnx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)



<!-- Sent from Reviewable.io -->

spiffyyeng commented:
> Cool. Did this survive with the webhook changes under `make stress`?
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67499#-:-MeUlDGW1qFw2B-eZG2H:b1zzdnx)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

yep! ran this numerous times with the concurrency changes.

spiffyyeng commented:
bors r+
